apiserver/facades/controller/crossmodelrelations/crossmodelrelations.go

```diff
@@ -69,7 +88,247 @@ func (api *CrossModelRelationsAPIv3) PublishRelationChanges(
+	relationDetails, err := api.relationService.GetRelationDetails(ctx, relationUUID)
+	if errors.Is(err, relationerrors.RelationNotFound) {
+		api.logger.Debugf(ctx, "relation %q not found when publishing relation changes", change.RelationToken)
+		return nil
+	} else if err != nil {
+		return err
+	} else if relationDetails.Life == life.Dead {
+		api.logger.Debugf(ctx, "relation %q is dead when publishing relation changes", change.RelationToken)
+		return nil
+	}
+
+	relationTag, err := constructRelationTag(relationDetails)
+	if err != nil {
+		return errors.Annotatef(err, "constructing relation tag for relation %q", relationUUID)
+	}
+
+	if err := api.checkMacaroonsForRelation(ctx, relationUUID, relationTag, change.Macaroons, change.BakeryVersion); err != nil {
+		return errors.Annotatef(err, "checking macaroons for relation %q", relationUUID)
+	}
+
+	// TODO (stickupkid): Work out if we get a offer UUID and get the
+	// application UUID from that instead.
+	applicationUUID, err := coreapplication.ParseID(change.ApplicationOrOfferToken)
+	if err != nil {
+		return err
+	}
+
+	// Ensure that the application is still alive.
+	appDetails, err := api.applicationService.GetApplicationDetails(ctx, applicationUUID)
+	if errors.Is(err, applicationerrors.ApplicationNotFound) || appDetails.Life == domainlife.Dead {
+		return errors.NotFoundf("application %q not found or dead when publishing relation changes for relation %q", applicationUUID, relationUUID)
+	} else if err != nil {
+		return err
+	}
+
+	switch {
+	case change.Life != life.Alive:
+		// We're dying or dead. We shouldn't continue onwards if we're dead,
+		// so return early.
+		if err := api.removalService.RemoveRemoteRelation(ctx, relationUUID); err != nil {
+			return errors.Annotatef(err, "removing remote relation %q", relationUUID)
+		}
+		if change.Life == life.Dead {
+			// No further processing for dead relations.
+			return nil
+		}
+
+	case change.Suspended != nil && *change.Suspended != relationDetails.Suspended:
+		if err := api.handleSuspendedRelationChange(ctx, relationUUID, *change.Suspended, change.SuspendedReason); err != nil {
+			return err
+		}
+	}
+
+	if err := api.handlePublishSettings(ctx, relationUUID, applicationUUID, appDetails.Name, change); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (api *CrossModelRelationsAPIv3) handleSuspendedRelationChange(
+	ctx context.Context,
+	relationUUID corerelation.UUID,
+	suspended bool,
+	suspendedReason string,
+) error {
+	var relationStatus status.Status
+	var message string
+	if suspended {
+		relationStatus = status.Suspended
+		message = suspendedReason
+	} else {
+		relationStatus = status.Joining
+		message = ""
+
+		// TODO (stickupkid): Check that the relation is still viable for
+		// consumption i.e. you can consume it again.
+	}
+
+	return api.statusService.SetRemoteRelationStatus(
+		ctx,
+		relationUUID,
+		status.StatusInfo{
+			Status:  relationStatus,
+			Message: message,
+		},
+	)
```

Comment:
I'll be doing a follow up to correctly handle the relation status here. It's not correct in either location.